File: test/system/TestBasisSet.py
import pulsar as psr

def run_test():
    tester = psr.PyTester("Testing the BasisSet and BasisSetShell")
    cGTO = psr.ShellType.CartesianGaussian
    sGTO = psr.ShellType.SphericalGaussian
    alpha=[3.42525091, 0.62391373, 0.16885540]
    c=[0.15432897, 0.53532814, 0.44463454]

    FakeD=psr.BasisShellInfo(cGTO,2,3,1,alpha,c)
    FakeD2=psr.BasisShellInfo(sGTO,2,3,1,alpha,c)
    carts=[0.0,0.0,0.0]
    H=psr.create_atom(carts,1)
    BI=psr.BasisInfo()
    BI.shells=[FakeD,FakeD2]
    H.basis_sets={"PRIMARY" :BI }
    GhH=psr.make_ghost_atom(H)
    Atoms=psr.AtomSetUniverse([H,GhH])
    Mol=psr.System(Atoms,True)
    BS=Mol.get_basis_set("PRIMARY")
    BS2=psr.BasisSet(BS)


    tester.test_equal("Copy constructors work",BS,BS2)
    BS3=psr.BasisSet(1,3,3,3)
    tester.test_return("Inequality works",True,True,BS3.__ne__,BS2)

    tester.test_return("Get types works",True,{cGTO,sGTO},BS.get_types)
    tester.test_return("Get n shells",True,4,BS2.n_shell)
    tester.test_return("Get n unique shells",True,2,BS.n_unique_shell)
    tester.test_return("Get n primitives",True,12,BS2.n_primitives)
    tester.test_return("Get n coeficients",True,12,BS2.n_coefficients)
    tester.test_return("Get number of functions",True,22,BS.n_functions)
    tester.test_return("Maximum number of primitivs",True,3,BS2.max_n_primitives)
    tester.test_return("Max angular momentum",True,2,BS2.max_am)
    tester.test_return("All angular momentum",True,{2},BS.all_am)
    tester.test_return("Max n functions in a shell",True,6,BS2.max_n_functions)
    tester.test_return("Shell start",True,6,BS2.shell_start,1)
    tester.test_call("Invalid shell start",False,BS2.shell_start,99)
    Si,Sj=BS.shell(3),BS.shell(2)
    tester.test_return("Shell has right coordinates",True,carts,Si.get_coords)
    tester.test_return("Shell has right coordinate",True,carts[1],Si.get_coord,1)
    tester.test_call("Get invalid shell",False,BS.shell,99)
    tester.test_return("Get unique shell",True,Si,BS.unique_shell,1)
    tester.test_call("Get invalid unique shell",False,BS.unique_shell,99)
    i=0
    for Sk in BS2:
       tester.test_equal("Iterator "+str(i),Sk,Si if i%2==1 else Sj)
       i=i+1

    tester.test_return("Get valid shell info",True,FakeD,BS2.shell_info,0)
    tester.test_call("Get invalid shell info",False,FakeD,BS.shell_info,99)
    BS4=psr.BasisSet(1,3,3,3)
    tester.test_return("Add shell that fits",True,None,BS4.add_shell,FakeD,carts)
    BS3.add_shell(FakeD,carts)
    tester.test_return("Shrink to fit",True,BS3,BS4.shrink_fit)
    tester.test_call("Add shell no fit",False,BS4.add_shell,FakeD2,carts)

    tester.test_return("Hash BS",True,BS.my_hash(),BS2.my_hash)

    tester.print_results()
    return tester.nfailed()

File: test/system/TestBasisSet.cpp
#include <pulsar/testing/CppTester.hpp>
#include <pulsar/system/System.hpp>

using namespace pulsar;
using ShellSet=std::set<ShellType>;

TEST_SIMPLE(TestBasisSet){

    ShellType cGTO = ShellType::CartesianGaussian;
    ShellType sGTO = ShellType::SphericalGaussian;
    //RMR- It is my understanding that BSS is really only going to be made by BS
    CppTester tester("Testing BasisSet and BasisSetShell");

    std::vector<double> alpha({3.42525091, 0.62391373, 0.16885540}),
    c({0.15432897, 0.53532814, 0.44463454});
    
    BasisShellInfo FakeD(cGTO, 2, 3, 1, alpha, c),
                  FakeD2(sGTO, 2, 3, 1, alpha, c);
    std::array<double,3> carts({0.0,0.0,0.0});
    Atom H=create_atom(carts,1);
    H.basis_sets["PRIMARY"].shells.push_back(FakeD);
    H.basis_sets["PRIMARY"].shells.push_back(FakeD2);
    Atom GhH=make_ghost_atom(H);
    AtomSetUniverse Atoms({H,GhH});
    System Mol(Atoms,true);
    BasisSet BS(std::move(Mol.get_basis_set("PRIMARY")));
    BasisSet BS2(BS);   
    tester.test_equal("Move and copy constructors work",BS,BS2);
    
    BasisSet BS3(2,6,6,6);
    tester.test_equal("Inequality works",true,BS3!=BS2);
    BS3=std::move(BS2);
    tester.test_equal("Move assignment works",BS3,BS);
    BS2=BS3;
    tester.test_equal("Copy assignment works",BS2,BS);
    
    //All basis sets up to BS3 should be equal at this point
    
    tester.test_member_return("Get types works",true,ShellSet({cGTO,sGTO}),
                              &BasisSet::get_types,&BS);
    tester.test_member_return("Get n shells",true,4,&BasisSet::n_shell,&BS2);
    tester.test_member_return("Get n unique shells",true,2,
                              &BasisSet::n_unique_shell,&BS);
    tester.test_member_return("Get n primitives",true,12,
                              &BasisSet::n_primitives,&BS3);
    tester.test_member_return("Get n coeficients",true,12,
                              &BasisSet::n_coefficients,&BS2);
    tester.test_member_return("Get number of functions",true,22,
                              &BasisSet::n_functions,&BS);
    tester.test_member_return("Maximum number of primitivs",true,3,
                              &BasisSet::max_n_primitives,&BS3);
    tester.test_member_return("Max angular momentum",true,2,&BasisSet::max_am,&BS2);
    std::set<int> am({2});
    tester.test_member_return("All angular momentum",true,am,&BasisSet::all_am,&BS);
    tester.test_member_return("Max n functions in a shell",true,6,
                              &BasisSet::max_n_functions,&BS3);
    tester.test_member_return("Shell start",true,6,&BasisSet::shell_start,&BS2,1);
    tester.test_member_call("Invalid shell start",false,&BasisSet::shell_start,
                            &BS2,99);
    const BasisSetShell& Si=BS.shell(3);
    const BasisSetShell& Sj=BS.shell(2);
    tester.test_member_return("Shell has right coordinates",true,carts,
                              &BasisSetShell::get_coords,&Si);
    tester.test_member_return("Shell has right coordinate",true,carts[1],
        &BasisSetShell::get_coord,&Si,1);
    tester.test_member_call("Get invalid shell",false,&BasisSet::shell,&BS,99);
    tester.test_member_call("Get unique shell",true,&BasisSet::unique_shell,&BS,1);
    tester.test_member_call("Get invalid unique shell",false,&BasisSet::unique_shell,&BS,99);
    size_t counter=0;
    for(const BasisSetShell& Sk: BS2)
    {
       tester.test_equal("Iterator "+std::to_string(counter),true, Sk==(counter%2?Si:Sj));
       ++counter;
    }
    
    
    tester.test_member_return("Get valid shell info",true,FakeD,
                              &BasisSet::shell_info,&BS3,0);
    tester.test_member_call("Get invalid shell info",false,
                            &BasisSet::shell_info,&BS3,99);
    BasisSet BS4(1,3,3,3),BS5(1,3,3,3);
    tester.test_member_call("Add shell that fits",true,&BasisSet::add_shell,
                            &BS4,FakeD,carts);
    BS5.add_shell(FakeD,carts);
    tester.test_member_return("Shrink to fit",true,BS5,&BasisSet::shrink_fit,&BS4);
    tester.test_member_call("Add shell no fit",false,&BasisSet::add_shell,&BS4,
                            FakeD2,carts);

    tester.test_equal("Hash BS",BS.my_hash(),BS2.my_hash());
    tester.test_equal("Hash BS2",BS.my_hash(),BS3.my_hash());
    
    tester.print_results();
    return tester.nfailed();
}

